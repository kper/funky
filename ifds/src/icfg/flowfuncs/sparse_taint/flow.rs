use crate::icfg::state::State;
use crate::icfg::{flowfuncs::*, tabulation::sparse::Ctx};

pub struct SparseTaintNormalFlowFunction;

impl SparseTaintNormalFlowFunction {
    pub fn identity<'a>(
        &self,
        ctx: &mut Ctx<'a>,
        function: &AstFunction,
        fact: &Fact,
        pc: usize,
        defuse: &mut DefUseChain,
        edges: &mut Vec<Edge>,
    ) -> Result<()> {
        let after = defuse.get_next(ctx, function, &fact.belongs_to_var, pc)?;

        for to in after {
            edges.push(Edge::Normal {
                from: fact.apply(),
                to: to,
                curved: false,
            });
        }

        Ok(())
    }
}

impl SparseNormalFlowFunction for SparseTaintNormalFlowFunction {
    fn flow<'a>(
        &self,
        ctx: &mut Ctx<'a>,
        function: &AstFunction,
        pc: usize,
        variable: &String,
        defuse: &mut DefUseChain,
    ) -> Result<Vec<Fact>> {
        debug!(
            "Calling flow for {} with var {} with pc {}",
            function.name, variable, pc
        );

        let mut facts = Vec::new();

        let instructions = &function.instructions;
        let instruction = instructions.get(pc);

        if instruction.is_none() {
            debug!("Instruction is none");
            return Ok(Vec::new());
        }

        let instruction = instruction.unwrap();

        debug!("Next instruction is {:?} for {}", instruction, variable);

        let mut nodes = defuse.demand(ctx, function, variable, pc)?;

        match instruction {
            Instruction::Unop(dest, ..) |
            Instruction::Phi(dest, ..) |
            Instruction::BinOp(dest, ..) | 
            Instruction::Assign(dest, ..) => {
                let x = defuse.demand_inclusive(ctx, function, dest, pc)?;
                nodes.extend(x.into_iter().map(|x| x.clone()));
            }
            _ => {}
        }
        let nodes = nodes
            .into_iter()
            .filter_map(|x| {
                if x.pc != x.next_pc {
                    Some(x.apply())
                } else {
                    Some(x)
                }
            })
            .collect::<Vec<_>>();

        facts.extend(nodes.into_iter());

        debug!("Out {:#?}", facts);

        Ok(facts)
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::icfg::state::State;
    use crate::ir::ast::Program;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_binop() {
        /*
            %0 = 1
            %1 = 1
            %2 = %0 op %1
            %3 = %1 op %0
        */

        let func_name = "main".to_string();
        let function = AstFunction {
            name: func_name.clone(),
            definitions: vec![
                "%0".to_string(),
                "%1".to_string(),
                "%2".to_string(),
                "%3".to_string(),
            ],
            instructions: vec![
                Instruction::Const("%0".to_string(), 1.0),
                Instruction::Const("%1".to_string(), 1.0),
                Instruction::BinOp("%2".to_string(), "%0".to_string(), "%1".to_string()),
                Instruction::BinOp("%3".to_string(), "%0".to_string(), "%1".to_string()),
            ],
            ..Default::default()
        };

        let mut graph = Graph::default();
        let mut state = State::default();

        let mut ctx = Ctx {
            graph: &mut graph,
            state: &mut state,
            prog: &Program {
                functions: vec![function.clone()],
            },
            block_resolver: HashMap::default(),
        };

        let pc = 0;

        // fullfilling precondition of `chain.cache()`
        ctx.state.init_function(&function, pc).unwrap();

        let mut defuse = DefUseChain::default();

        let sparse = SparseTaintNormalFlowFunction;

        let facts = sparse
            .flow(&mut ctx, &function, 2, &"%0".to_string(), &mut defuse)
            .unwrap();

        debug!("facts {:#?}", facts);
        assert_eq!(4, facts.len());

        let cmp_facts: Vec<Fact> = vec![
            Fact {
                belongs_to_var: "%0".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 4,
                next_pc: 4,
                track: 1,
                function: "main".to_string(),
                memory_offset: None,
            },
            Fact {
                belongs_to_var: "%0".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 4,
                next_pc: 4,
                track: 1,
                function: "main".to_string(),
                memory_offset: None,
            },
            Fact {
                belongs_to_var: "%2".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 3,
                next_pc: 4,
                track: 3,
                function: "main".to_string(),
                memory_offset: None,
            },
            Fact {
                belongs_to_var: "%2".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 4,
                next_pc: 4,
                track: 3,
                function: "main".to_string(),
                memory_offset: None,
            },
        ];
        assert_eq!(cmp_facts, facts);
    }
}
