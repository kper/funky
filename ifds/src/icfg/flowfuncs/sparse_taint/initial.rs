use crate::icfg::tabulation::sparse::defuse::DefUseChain;
use crate::icfg::{flowfuncs::*, tabulation::sparse::Ctx};

pub struct SparseTaintInitialFlowFunction;

impl SparseInitialFlowFunction for SparseTaintInitialFlowFunction {
    fn flow<'a>(
        &self,
        ctx: &mut Ctx<'a>,
        function: &AstFunction,
        pc: usize,
        init_facts: &Vec<Fact>,
        defuse: &mut DefUseChain,
    ) -> Result<Vec<Edge>> {
        debug!("Calling init flow for {} with pc {}", function.name, pc);

        let mut edges = Vec::new();

        // We need the offset, because not every
        // instruction is taintable. For example BLOCK and JUMP
        // have no registers. That's why skip to the next one.
        let offset = 0;
        let instructions = &function.instructions;

        let init_fact = init_facts.get(0).context("Cannot find taut")?.clone();

        debug!("init fact is {:#?}", init_fact);

        loop {
            let pc = pc + offset;
            let instruction = instructions.get(pc).context("Cannot find instr")?;
            debug!("Next instruction is {:?}", instruction);

            match instruction {
                Instruction::Const(dest, _)
                | Instruction::Unknown(dest)
                | Instruction::Assign(dest, _)
                | Instruction::Unop(dest, _)
                | Instruction::BinOp(dest, _, _)
                | Instruction::Phi(dest, _, _)
                | Instruction::Conditional(dest, _)
                | Instruction::Load(dest, _, _) => {
                    let before = vec![init_fact.clone()];

                    for b in before.into_iter() {
                        let after_taut = defuse
                            .demand(ctx, &function, &"taut".to_string(), b.pc)
                            .context("Cannot find taut's fact")?;

                        debug!("Next taut is {:#?}", after_taut);

                        for taut in after_taut.into_iter() {
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: taut.clone(),
                            });
                        }

                        let after_var = defuse
                            .demand(ctx, &function, dest, pc)
                            .context("Cannot find var's fact")?
                            .into_iter()
                            .map(|x| x.clone())
                            .collect::<Vec<_>>();

                        // append all left sides to the nodes
                        // %2 = binop %0 %1 -- there %2 is the left side
                        let mut appended = Vec::new();
                        let mut append_lhs = |dest: &String| -> Result<()> {
                            defuse.force_remove_if_outdated(function, dest, pc)?;
                            let x = defuse.demand_inclusive(ctx, function, dest, pc)?;
                            appended.extend(x.into_iter().map(|x| x.clone()));

                            Ok(())
                        };

                        for var in after_var.clone().iter() {
                            append_lhs(&var.belongs_to_var)?;
                        }

                        for var in after_var.into_iter().chain(appended) {
                            let applied = var.apply();

                            // after
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: applied.clone(),
                            });
                        }
                    }
                }
                Instruction::Call(_callee, _, dests) => {
                    // When the analysis starts at a Call Instruction,
                    // then the function doesn't get called.
                    // But all destinations get automatically tainted.

                    let before = vec![init_fact.clone()];
                    for b in before.into_iter() {
                        let after_taut = defuse
                            .demand(ctx, &function, &"taut".to_string(), b.pc)
                            .context("Cannot find taut's fact")?;

                        debug!("Next taut is {:#?}", after_taut);

                        for taut in after_taut.into_iter() {
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: taut.clone(),
                            });
                        }
                    }
                    for dest in dests {
                        let after_var = defuse
                            .demand(ctx, &function, dest, pc)
                            .context("Cannot find var's fact")?
                            .into_iter()
                            .map(|x| x.clone())
                            .collect::<Vec<_>>();

                        // append all left sides to the nodes
                        // %2 = binop %0 %1 -- there %2 is the left side
                        let mut appended = Vec::new();
                        let mut append_lhs = |dest: &String| -> Result<()> {
                            defuse.force_remove_if_outdated(function, dest, pc)?;
                            let x = defuse.demand_inclusive(ctx, function, dest, pc)?;
                            appended.extend(x.into_iter().map(|x| x.clone()));

                            Ok(())
                        };

                        for var in after_var.clone().iter() {
                            append_lhs(&var.belongs_to_var)?;
                        }

                        for var in after_var.into_iter().chain(appended) {
                            let applied = var.apply();

                            // after
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: applied.clone(),
                            });
                        }
                    }
                }
                Instruction::Store(_src, offset, _) => {
                    let before = vec![init_fact.clone()];

                    for b in before.into_iter() {
                        // The tautological fact was built by the `pacemaker`
                        // and will not be sparsely propagated.

                        let after_taut = defuse
                            .demand(ctx, &function, &"taut".to_string(), pc)
                            .context("Cannot find taut's fact")?;

                        for taut in after_taut.into_iter() {
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: taut.clone(),
                            });
                        }

                        let mem = ctx
                            .state
                            .add_memory_var(function.name.clone(), offset.clone() as usize);

                        let after_var = defuse
                            .demand_inclusive(ctx, &function, &mem.name, pc)
                            .context("Cannot find var's fact")?;

                        for var in after_var.into_iter() {
                            let mut applied = var.apply();

                            assert!(applied.pc <= applied.next_pc);

                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: applied.clone(),
                            });
                        }
                    }
                }
                Instruction::Return(_dests) => {
                    let before = vec![init_fact.clone()];

                    for b in before.into_iter() {
                        // The tautological fact was built by the `pacemaker`
                        // and will not be sparsely propagated.

                        let after_taut = defuse
                            .demand(ctx, &function, &"taut".to_string(), pc)
                            .context("Cannot find taut's fact")?;

                        for taut in after_taut.into_iter() {
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: taut.clone(),
                            });
                        }

                        let after_var = defuse
                            .demand(ctx, &function, &"taut".to_string(), pc)
                            .context("Cannot find var's fact")?;

                        for var in after_var.into_iter() {
                            let applied = var.apply();

                            assert!(applied.pc <= applied.next_pc);

                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: var.clone(),
                            });
                        }
                    }
                }
                Instruction::Block(_) | Instruction::Jump(_) => {
                    panic!("Block or Jump as first instruction is not supported");
                }
                _ => {}
            }

            break;
        }

        Ok(edges)
    }
}
