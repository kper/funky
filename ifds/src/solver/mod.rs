//! The implementation for extracting the taints from the graph.

use crate::icfg::graph::{Edge, Graph};
use anyhow::{Context, Result};
use std::collections::HashSet;

type PC = usize;

pub struct IfdsSolver;

/// Represents the taints from the execution of the solver.
#[derive(Debug)]
pub struct Taint {
    pub variable: String,
    pub function: String,
    pub pc: PC,
}

/// Request is the datastructure which defines at what instruction the [`Graph`] should be build from.
#[derive(Debug, Clone)]
pub struct Request {
    pub variable: Option<String>,
    pub function: String,
    pub pc: PC,
}

pub trait Solver {
    /// Get all taints for the graph. The edges must be already computed.
    /// The result depends on the `req` [`Request`]. Only the requested function and instruction which have
    /// an higher or equal program counter.
    fn all_sinks(&mut self, graph: &mut Graph, req: &Request) -> Result<Vec<Taint>>;

    /// Check if the statement at `req` and the statement `resp` have a taint.
    fn is_taint(&mut self, graph: &mut Graph, req: &Request, resp: &Request) -> Result<bool>;

    /// Return all tainted variables names at any statement for the graph. The edges must be already computed.
    /// The result depends on the `req` [`Request`]. Only the requested function and instruction which have
    /// an higher or equal program counter.
    fn sinks_var(&mut self, graph: &mut Graph, req: &Request) -> Result<HashSet<String>>;
}

impl Solver for IfdsSolver {
    fn all_sinks(&mut self, graph: &mut Graph, req: &Request) -> Result<Vec<Taint>> {
        assert!(req.variable.is_some());
        assert!(req.variable.as_ref().unwrap().starts_with("%"));

        let function = &req.function;

        let taints: Vec<_> = graph
            .edges
            .iter()
            .filter(|x| {
                matches!(x, Edge::Path { .. })
                    && &x.to().function == function
                    //&& x.get_from().var_is_taut
                    && x.get_from().next_pc == req.pc
            })
            .map(|x| x.to())
            .filter(|x| !x.var_is_taut)
            .map(|x| Taint {
                function: x.function.clone(),
                pc: x.next_pc.checked_sub(1).unwrap_or(0),
                variable: x.belongs_to_var.clone(),
            })
            .collect();

        Ok(taints)
    }

    fn sinks_var(&mut self, graph: &mut Graph, req: &Request) -> Result<HashSet<String>> {
        assert!(req.variable.is_some());
        assert!(req.variable.as_ref().unwrap().starts_with("%"));

        let function = &req.function;

        let taints = graph
            .edges
            .iter()
            .filter(|x| {
                matches!(x, Edge::Path { .. })
                    && &x.to().function == function
                    && x.get_from().next_pc == req.pc
            })
            .map(|x| x.to())
            .filter(|x| !x.var_is_taut)
            .map(|x| x.belongs_to_var.clone())
            .collect::<HashSet<_>>();

        Ok(taints)
    }

    fn is_taint(&mut self, graph: &mut Graph, req: &Request, resp: &Request) -> Result<bool> {
        let all_sinks = self
            .all_sinks(graph, req)
            .context("Cannot compute the sinks")?;

        let var = resp
            .variable
            .as_ref()
            .context("Please specify a variable for the response")?;

        Ok(all_sinks
            .into_iter()
            .find(|x| x.function == resp.function && x.pc == resp.pc && &x.variable == var)
            .is_some())
    }
}

pub trait GraphReachability {
    /// Get all taints for the graph. The edges must be already computed.
    /// The result depends on the `req` [`Request`]. Only the requested function and instruction which have
    /// an higher or equal program counter.
    fn all_sinks(&mut self, graph: &mut Graph, req: &Request) -> Vec<Taint>;
}
