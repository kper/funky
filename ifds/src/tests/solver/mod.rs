use crate::solver::*;

use crate::icfg::convert::ConvertSummary;
use crate::icfg::tikz::render_to;
use insta::assert_snapshot;

use crate::grammar::*;

use crate::icfg::flowfuncs::taint::flow::TaintNormalFlowFunction;
use crate::icfg::flowfuncs::taint::initial::TaintInitialFlowFunction;

macro_rules! ir {
    ($name:expr, $req:expr, $ir:expr) => {{
        let mut convert = ConvertSummary::new(TaintInitialFlowFunction, TaintNormalFlowFunction);

        let prog = ProgramParser::new().parse(&$ir).unwrap();

        let (graph, state) = convert.visit(&prog, &$req).unwrap();

        let output = render_to(&graph, &state);

        assert_snapshot!(format!("{}_dot", $name), output);

        graph
    }};
}

macro_rules! check_vars {
    ($name:expr, $solver:expr, $graph:expr, $req:expr) => {
        let mut vars = $solver
                    .sinks_var(&mut $graph, &$req)
                    .unwrap()
                    .into_iter()
                    .collect::<Vec<_>>();

        vars.sort();

        assert_snapshot!(
            format!("{}_vars", $name),
            format!(
                "{:#?}",
                vars 
            )
        );
    };
}

#[test]
fn test_intra_reachability() {
    let mut solver = IfdsSolver;

    let name = "intra_reach";

    let req = Request {
        function: "test".to_string(),
        pc: 0,
        variable: Some("%0".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define test (result 0) (define %0 %1) {
            %0 = 1
            %1 = %0
         };
    "
    );

    let req = Request {
        variable: Some("%0".to_string()),
        function: "test".to_string(),
        pc: 0,
    };
    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 0,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 1,
                variable: Some("%1".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 1,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_kill() {
    let mut solver = IfdsSolver;

    let name = "test_kill";

    let req = Request {
        function: "test".to_string(),
        pc: 0,
        variable: Some("%0".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define test (result 0) (define %0) {
            %0 = 1
            KILL %0
         };
    "
    );

    let req = Request {
        variable: Some("%0".to_string()),
        function: "test".to_string(),
        pc: 0,
    };
    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 0,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    // should not be reachable anymore
    assert!(!solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 1,
                variable: Some("%1".to_string()),
            }
        )
        .unwrap());

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_loop() {
    let mut solver = IfdsSolver;

    let name = "test_loop";

    let req = Request {
        function: "main".to_string(),
        pc: 1,
        variable: Some("%0".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define main (result 0) (define %0 %1) {
            BLOCK 0
            %0 = 1
            GOTO 0 
        };
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "main".to_string(),
                pc: 1,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_functions() {
    let mut solver = IfdsSolver;

    let name = "test_functions";

    let req = Request {
        function: "test".to_string(),
        pc: 0,
        variable: Some("%0".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define test (result 0) (define %0) {
            %0 = 1
            CALL mytest(%0)
        };
        define mytest (param %0) (result 0) (define %0 %1)  {
            %0 = 2   
            %1 = 3
            RETURN;
        };
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 0,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 1,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_gcd() {
    let mut solver = IfdsSolver;

    let name = "test_gcd";

    let req = Request {
        function: "0".to_string(),
        pc: 17,
        variable: Some("%11".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
       define 0 (param %0 %1) (result 1) (define %0 %1 %2 %3 %4 %5 %6 %7 %8 %9 %10 %11 %12 %13 %14 %15 %16) {
        BLOCK 0
        BLOCK 1
        %4 = %0
        %5 = %1
        %6 = %5 op %4
        IF %6 THEN GOTO 3 ELSE GOTO 4
        BLOCK 3 
        %7 = %0
        %2 = %7
        KILL %7
        GOTO 2
        GOTO 4
        BLOCK 4
        BLOCK 5
        %8 = %1
        %9 = 0
        %10 = %0
        %11 = %1
        %12 = %11 op %10
        %13 = %12
        %2 = %12
        %14 = %13 op %12
        %3 = %14
        %15 = %1
        %16 = 0
        BLOCK 2
        RETURN %15;
        }; 
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_globals() {
    let mut solver = IfdsSolver;

    let name = "test_globals";

    let req = Request {
        function: "test".to_string(),
        pc: 0,
        variable: Some("%0".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define test (result 0) (define %-1 %0 %2) {
            %0 = 1
            %-1 = %0 
            %2 <- CALL mytest()
        };
        define mytest (param) (result 1) (define %-1 %0 %1)  {
            %0 = 2   
            %1 = 3
            RETURN %-1;
        };
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 0,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 1,
                variable: Some("%-1".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 2,
                variable: Some("%2".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 2,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_returned_value() {
    let mut solver = IfdsSolver;

    let name = "test_returned_values";

    let req = Request {
        function: "0".to_string(),
        pc: 8,
        variable: Some("%11".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define 0 (param %0) (result 1) (define %0 %1 %7 %8 %9 %10 %11 %12 %13 %14 %15 %16 %17) {
            BLOCK 0
            BLOCK 1
            BLOCK 3
            %1 = %0
            %7 = %0
            %8 = -1
            %9 = %8 op %7
            %10 <- CALL 0(%9)
            %11 = %0
            %12 = -2
            %13 = %12 op %11
            %14 <- CALL 0(%13)
            %15 = %14 op %13
            KILL %15
            KILL %14
            RETURN %0;
        }; 
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_looped_param() {
    let mut solver = IfdsSolver;

    let name = "test_looped_param";

    let req = Request {
        function: "0".to_string(),
        pc: 1,
        variable: Some("%7".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define 0 (param %0) (result 1) (define %0 %1 %2 %3 %4 %5 %6 %7 %8 %9 %10 %11 %12 %13 %14 %15 %16 %17) {
            %1 = %0
            %7 = %0
            %9 = %8 op %7
            %10 <- CALL 0(%9)
            %11 = %0
            RETURN %0;
        };
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_memory() {
    let mut solver = IfdsSolver;

    let name = "test_memory";

    let req = Request {
        function: "0".to_string(),
        pc: 0,
        variable: Some("%0".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define 0 (param %0) (result 0) (define %0 %1 %2 %3 %4 %5 %6 %7) {
            %0 = -12345
            STORE FROM %0 OFFSET 0 + %0 ALIGN 2 32
            %4 = 8
            %5 = LOAD OFFSET 0 + %4 ALIGN 0
            %6 = 8
            %7 = LOAD OFFSET 0 + %6 ALIGN 0
        };
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "0".to_string(),
                pc: 0,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "0".to_string(),
                pc: 1,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "0".to_string(),
                pc: 1,
                variable: Some("mem@0".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "0".to_string(),
                pc: 3,
                variable: Some("%5".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "0".to_string(),
                pc: 5,
                variable: Some("%7".to_string()),
            }
        )
        .unwrap());

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_ir_multiple_functions() {
    let mut solver = IfdsSolver;

    let name = "test_multiple_functions";

    let req = Request {
        function: "test".to_string(),
        pc: 0,
        variable: Some("%0".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 <- CALL mytest(%0)
            %2 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0)  {
            RETURN %0;
        };
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 1,
                variable: Some("%1".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 2,
                variable: Some("%2".to_string()),
            }
        )
        .unwrap());

    check_vars!(name, solver, graph, req);
}
