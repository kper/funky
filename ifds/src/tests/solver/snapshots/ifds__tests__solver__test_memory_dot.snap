---
source: src/tests/solver/mod.rs
expression: output

---
   \documentclass{standalone}
            \usepackage{pgf, tikz}
            \usetikzlibrary{arrows, automata}
            \begin{document}
                \begin{tikzpicture}
                    [node distance=3cm]
                    \tikzstyle{every state}=[
                        draw = black,
                        thick,
                        fill = white,
                        minimum size = 4mm
                    ]

                    \node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (0) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (1) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (2) at (0, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (3) at (0, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (4) at (0, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (5) at (0, 5) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (6) at (0, 6) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (7) at (4, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $@0$}] (8) at (1, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (9) at (4, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $@0$}] (10) at (1, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (11) at (4, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%5$}] (12) at (3, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $@0$}] (13) at (1, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (14) at (4, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%5$}] (15) at (3, 5) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $@0$}] (16) at (1, 5) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (17) at (4, 5) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%5$}] (18) at (3, 6) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%7$}] (19) at (2, 6) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $@0$}] (20) at (1, 6) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (21) at (4, 6) { };
\node[font=\tiny] (note_0) at (-1.5, 0.5) {Const(0, -12345.0)};
\node[font=\tiny] (note_1) at (-1.5, 1.5) {Store(0, 0.0, 0)};
\node[font=\tiny] (note_2) at (-1.5, 2.5) {Const(4, 8.0)};
\node[font=\tiny] (note_3) at (-1.5, 3.5) {Load(5, 0.0, 4)};
\node[font=\tiny] (note_4) at (-1.5, 4.5) {Const(6, 8.0)};
\node[font=\tiny] (note_5) at (-1.5, 5.5) {Load(7, 0.0, 6)};
		\path[->, blue] (0) [loop right] edge  node { } (0);
		\path[->, blue] (0) [bend right] edge  node { } (1);
		\path[->, blue] (0) [bend right] edge  node { } (2);
		\path[->, blue] (0) [bend right] edge  node { } (3);
		\path[->, blue] (0) [bend right] edge  node { } (4);
		\path[->, blue] (0) [bend right] edge  node { } (5);
		\path[->, blue] (0) [bend right] edge  node { } (6);
		\path[->, blue] (0) [bend right] edge  node { } (7);
		\path[->, blue] (0) [bend right] edge  node { } (8);
		\path[->, blue] (0) [bend right] edge  node { } (9);
		\path[->, blue] (0) [bend right] edge  node { } (10);
		\path[->, blue] (0) [bend right] edge  node { } (11);
		\path[->, blue] (0) [bend right] edge  node { } (12);
		\path[->, blue] (0) [bend right] edge  node { } (13);
		\path[->, blue] (0) [bend right] edge  node { } (14);
		\path[->, blue] (0) [bend right] edge  node { } (15);
		\path[->, blue] (0) [bend right] edge  node { } (16);
		\path[->, blue] (0) [bend right] edge  node { } (17);
		\path[->, blue] (0) [bend right] edge  node { } (18);
		\path[->, blue] (0) [bend right] edge  node { } (19);
		\path[->, blue] (0) [bend right] edge  node { } (20);
		\path[->, blue] (0) [bend right] edge  node { } (21);


                \end{tikzpicture}
            \end{document}
